osu.Game.Rulesets.Osu/Difficulty/Skills/Rhythm.cs

```diff
@@ -0,0 +1,136 @@
+                        }
+                        else
+                        {
+                            if (Previous[i - 1].BaseObject is Slider) // bpm change is into slider, this is easy acc window
+                                effectiveRatio *= 0.125;
+
+                            if (Previous[i].BaseObject is Slider) // bpm change was from a slider, this is easier typically than circle -> circle
+                                effectiveRatio *= 0.25;
+
+                            if (previousIslandSize == islandSize) // repeated island size (ex: triplet -> triplet)
+                                effectiveRatio *= 0.25;
+
+                            if (previousIslandSize % 2 == islandSize % 2) // repeated island polartiy (2 -> 4, 3 -> 5)
+                                effectiveRatio *= 0.50;
+
+                            if (lastDelta > prevDelta + 10 && prevDelta > currDelta + 10) // previous increase happened a note ago, 1/1->1/2-1/4, dont want to buff this.
+                                effectiveRatio *= 0.125;
+
+                            rhythmComplexitySum += Math.Sqrt(effectiveRatio * startRatio) * currHistoricalDecay * Math.Sqrt(4 + islandSize) / 2 * Math.Sqrt(4 + previousIslandSize) / 2;
+
+                            startRatio = effectiveRatio;
+
+                            previousIslandSize = islandSize; // log the last island size.
+
+                            if (prevDelta * 1.25 < currDelta) // we're slowing down, stop counting
+                                firstDeltaSwitch = false; // if we're speeding up, this stays true and  we keep counting island size.
+
+                            islandSize = 1;
+                        }
+                    }
+                    else if (prevDelta > 1.25 * currDelta) // we want to be speeding up.
+                    {
+                        // Begin counting island until we change speed again.
+                        firstDeltaSwitch = true;
+                        startRatio = effectiveRatio;
+                        islandSize = 1;
+                    }
+                }
+            }
+
+            return Math.Sqrt(4 + rhythmComplexitySum * rhythm_multiplier) / 2; // produces multiplier that can be applied to strain. range [1, infinity) (not really though)
+        }
+
+        private double strainDecay(double ms) => Math.Pow(strainDecayBase, ms / 1000);
+
+        protected override double CalculateInitialStrain(double time) => currentStrain * strainDecay(time - Previous[0].StartTime);
+
+        protected override double StrainValueAt(DifficultyHitObject current)
+        {
+            double rhythmBonus = calculateRhythmBonus(current);
+            currentStrain *= strainDecay(current.DeltaTime);
+            currentStrain += rhythmBonus - 1;
+
+            if (current is OsuDifficultyHitObject osuCurrObj)
+                osuCurrObj.RhythmDifficulty = rhythmBonus;
```

Comment:
Not sure about this...

Have you read my thoughts on this at https://github.com/ppy/osu/discussions/16122#discussioncomment-1861738?

Reply:
I'm not sure about this either honestly. It's structured that way only to keep values from changing because of how rhythm bonus was done in the first place.
It requires both exposing every calculated strain to `Speed` and access to `Skill`'s strain summing and object history, so moving it out into a standalone helper isn't really an option

Reply:
Storing in ODHO makes sense to me personally, seems intuitive to me at least. Rhythm should stay as a skill though, especially the direction forward is that it's going to be aggregated and used for https://github.com/ppy/osu/pull/15091 AFAIK

Reply:
So, although it's not being used here, the idea is that you may want to refer not just to the current object's rhythm strain but a past object's rhythm strain?

Reply:
Hmm, in principle I can't see why not but I'm not immediately seeing a use case for that. I mentioned storing in ODHO making sense in terms of carrying data from one skill to another and I hadn't even considered doing something like `Previous[0].RhythmDifficulty`. 

That's besides the point, I have read your comment in https://github.com/ppy/osu/discussions/16122#discussioncomment-1953131 and I really like the direction proposed here, so that has my full support.